src/fairchem/core/common/relaxation/optimizers/lbfgs_torch.py

```diff
@@ -108,29 +90,30 @@ def run(self, fmax, steps):
             ]
 
         iteration = 0
-        converged = False
-        converged_mask = torch.zeros_like(
-            self.batch.atomic_numbers, device=self.device
-        ).bool()
-        while iteration < steps and not converged:
-            _converged_mask, energy, forces = self.check_convergence(iteration)
-            # Models like GemNet-OC can have random noise in their predictions.
-            # Here we ensure atom positions are not being updated after already
-            # hitting the desired convergence criteria.
-            converged_mask = torch.logical_or(converged_mask, _converged_mask)
-            converged = torch.all(converged_mask)
-            update_mask = torch.logical_not(converged_mask)
-
-            if self.trajectories is not None and (
-                self.save_full or converged or iteration == steps - 1 or iteration == 0
-            ):
-                self.write(energy, forces, update_mask)
-
-            if not converged and iteration < steps - 1:
-                self.step(iteration, forces, update_mask)
+        max_forces = self.optimizable.get_max_forces(apply_constraint=True)
+        logging.info("Step   Fmax(eV/A)")
+        while iteration < steps and not self.optimizable.converged(
+            forces=None, fmax=self.fmax, max_forces=max_forces
+        ):
+            logging.info(
+                f"{iteration} " + " ".join(f"{x:0.3f}" for x in max_forces.tolist())
+            )
+
+            if self.trajectories is not None and self.save_full is True:
+                self.write()
```

Comment:
FYI - `save_full=False` should save the initial and the final frame. This logic would prevent saving the initial frame in that scenario.